### DISCUSSION TITLE
How do I mock dependency outputs within hooks in my terragrunt.hcl configuration?

### DISCUSSION THREAD

**Discussion by vishvikkrishnan** [gruntwork-io/knowledge-base#812] [How do I]
I use a `before_hook` which executes a command with a "dynamic" value in my terragrunt.hcl file. The block is able to read values from the locals block from the root configuration, however, it fails to resolve the dependency output. Here is the before_hook block:
```hcl
before_hook "app_login" {
    commands = ["apply", "destroy"]
    execute = [
      "app", "login", include.global.locals.app_admin_details.username, include.global.locals.app_url,
      "${chomp(dependency.app_install.outputs.app_api_key.content)}"
    ]
  }
```
The above block executes just fine for operations like `terragrunt apply` and `terragrunt destroy` because the outputs are already available in the state for terragrunt to fetch. However, other operations like init, validate, validate-inputs, and plan result in an error.
```shell
ERRO[0003]   on /home/ubuntu/deployment/terragrunt/site_build/app/app_config/terragrunt.hcl line 6, in terraform: 
ERRO[0003]    6:     execute = [              
ERRO[0003]    7:       "app", "login", include.global.locals.app_admin_details.username, include.global.locals.app_url, 
ERRO[0003]    8:       "${chomp(dependency.app_install.outputs.app_api_key.content)}" 
ERRO[0003]    9:     ]                                  
ERRO[0003]                                              
ERRO[0003] Unsuitable value: value must be known
```

I'd like to know how to "mock" such outputs that are computed from module dependencies inside the hooks so the other operations continue to work. Happy to open a bug/feature request if this is a valid use case.

---

<ins datetime="2024-03-28T22:50:59Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110850">Tracked in ticket #110850</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
was considered to use "mock_outputs_*" configuration keys? , like:

```

dependency "dep" {
  config_path = "../dep"
  mock_outputs_allowed_terraform_commands = ["validate", "init"]
  mock_outputs = {
    data = "fake-value"
  }
}

terraform {
  source = "."

  before_hook "app_login" {
    commands = ["apply", "destroy"]
    execute = [
      "echo", "test output:", dependency.dep.outputs.data
    ]
  }
}
```

https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#dependency

Simplified example in: https://github.com/denis256/terragrunt-tests/tree/master/discussion-812
